modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js

```diff
@@ -251,8 +203,78 @@ export default class Tile3DLayer extends CompositeLayer {
     );
   }
 
+  _create3DModelTileLayer(tileHeader) {
+    const {gltf, instances, cartographicOrigin, modelMatrix} = tileHeader.content;
+
+    const SubLayerClass = this.getSubLayerClass('scenegraph', ScenegraphLayer);
+
+    return new SubLayerClass(
+      {
+        _lighting: 'pbr'
+      },
+      this.getSubLayerProps({
+        id: 'scenegraph'
+      }),
+      {
+        id: `${this.id}-scenegraph-${tileHeader.id}`,
+        // Fix for ScenegraphLayer.modelMatrix, under flag in deck 7.3 to avoid breaking existing code
+        data: instances || [{}],
+        scenegraph: gltf,
+
+        coordinateSystem: COORDINATE_SYSTEM.METER_OFFSETS,
+        coordinateOrigin: cartographicOrigin,
+        modelMatrix,
+        getTransformMatrix: instance => instance.modelMatrix,
+        getPosition: instance => [0, 0, 0]
+      }
+    );
+  }
+
+  _createSimpleMeshLayer(tileHeader) {
+    const content = tileHeader.content;
+    const {attributes, modelMatrix, cartographicOrigin, texture} = content;
+    const {normals, texCoords} = attributes;
+    const positions = new Float32Array(attributes.positions.value.length);
+    for (let i = 0; i < positions.length; i += 3) {
+      scratchOffset.copy(modelMatrix.transform(attributes.positions.value.subarray(i, i + 3)));
+      positions.set(scratchOffset, i);
+    }
+
+    const geometry = new Geometry({
+      drawMode: GL.TRIANGLES,
+      attributes: {
+        positions,
+        normals,
+        texCoords
+      }
+    });
+
+    const SubLayerClass = this.getSubLayerClass('simplemesh', SimpleMeshLayer);
+
+    return new SubLayerClass(
+      this.getSubLayerProps({
+        id: 'simplemesh'
+      }),
+      {
+        id: `${this.id}-simplemesh-${tileHeader.id}`,
+        mesh: geometry,
+        data: [{}],
+        getPosition: [0, 0, 0],
+        getColor: [255, 255, 255],
+        texture,
+        coordinateOrigin: cartographicOrigin,
+        coordinateSystem: COORDINATE_SYSTEM.METER_OFFSETS
+      }
+    );
+  }
+
   renderLayers() {
-    return this.state.layers;
+    const {tileset3d, layerMap} = this.state;
+    if (!tileset3d) {
+      return null;
+    }
+    this._updateLayerMap();
```

Comment:
Too many intermediate arrays are created here. Directly call `tileset3d.tiles.map` similar to the 2d tiles layer.